mhs/mhs/outbound/request/synchronous/handler.py

```diff
@@ -1,91 +1,108 @@
-                                                                                            self.request.body.decode(),
-                                                                                            message_id)
-
-            if interaction_is_async:
-                self.callbacks[async_message_id] = self._write_async_response
-                logger.info('0002', 'Added callback for asynchronous message with {MessageId}',
-                            {'MessageId': async_message_id})
-
-            immediate_response = self.workflow.send_message(interaction_name, message)
-
-            if interaction_is_async:
-                await self._pause_request(async_message_id)
-            else:
-                # No async response expected. Just return the response to our initial request.
-                self._write_response(immediate_response)
-
-        except common_workflow.UnknownInteractionError:
-            raise tornado.web.HTTPError(404, "Unknown interaction ID: %s", interaction_name)
-        finally:
-            if async_message_id in self.callbacks:
-                del self.callbacks[async_message_id]
-
-    async def _pause_request(self, async_message_id):
-        """Pause the incoming request until an asynchronous response is received (or a timeout is hit).
-
-        :param async_message_id: The ID of the request that expects an asynchronous response.
-        :raises: An HTTPError if we timed out waiting for an asynchronous response.
-        """
+            interaction_details = self._get_interaction_details(interaction_id)
+        except common_workflow.UnknownInteractionError as e:
+            logger.warning('0007', 'Unknown {InteractionId} in request', {'InteractionId': interaction_id})
+            raise tornado.web.HTTPError(404, "Unknown interaction ID: %s", interaction_id) from e
+
         try:
-            logger.info('0003', 'Waiting for asynchronous response to message with {MessageId}',
-                        {'MessageId': async_message_id})
-            await self.async_response_received.wait(datetime.timedelta(seconds=self.async_timeout))
-        except TimeoutError:
-            raise tornado.web.HTTPError(log_message=f"Timed out waiting for a response to message {async_message_id}")
+            workflow = self.workflows[interaction_details['workflow']]
+        except KeyError as e:
+            logger.error('0008', "Weren't able to determine workflow for {InteractionId} . This likely is due to a "
+                                 "misconfiguration in interactions.json", {"InteractionId": interaction_id})
+            raise tornado.web.HTTPError(500, "Couldn't determine workflow to invoke for interaction ID: %s",
+                                        interaction_id) from e
+
+        status, response = await workflow.handle_supplier_message(message_id, interaction_details, body)
+
+        self._write_response(status, response)
```

Comment:
So this might be something worth double checking when Dave gets back but I recall him mentioning that we should be recording message information almost as soon as we receive it in the state store. Not sure if this wants to be done in the handler or when it enters a work flow though

Reply:
As discussed this will be done as soon as the message enters the workflow